File: doc-examples/mongo-example-kotlin-ksp/src/main/kotlin/example/QuantityAttributeConverter.kt
package example

import io.micronaut.core.convert.ConversionContext
import io.micronaut.data.model.runtime.convert.AttributeConverter
import jakarta.inject.Singleton

@Singleton // <1>
class QuantityAttributeConverter : AttributeConverter<Quantity?, Int?> {

    // <2>
    override fun convertToPersistedValue(quantity: Quantity?, context: ConversionContext): Int? {
        return quantity?.amount
    }

    // <3>
    override fun convertToEntityValue(value: Int?, context: ConversionContext): Quantity? {
        return if (value == null) null else Quantity(value)
    }

}

File: doc-examples/jdbc-example-python/src/main/python/example/QuantityAttributeConverter.py
from typing import TYPE_CHECKING

from jakarta.inject import Singleton
from micronaut.core.convert import ConversionContext
from micronaut.data.model.runtime.convert import AttributeConverter

if TYPE_CHECKING:
    from example.Quantity import Quantity


@Singleton  # <1>
class QuantityAttributeConverter(AttributeConverter["Quantity", int]):

    def convertToPersistedValue(self, quantity: "Quantity | None", context: ConversionContext) -> int | None:  # <2>
        return None if quantity is None else quantity.amount

    def convertToEntityValue(self, value: int | None, context: ConversionContext) -> "Quantity | None":  # <3>
        from example.Quantity import Quantity
        return None if value is None else Quantity(value)
